syntax = "proto3";

package Protocol;

// ==========================================================
// Packet IDs (MsgId)
// ==========================================================
enum MsgId {
  NONE = 0;
  
  // Auth & Room (100-199)
  C_LOGIN = 100;
  S_LOGIN = 101;
  C_CREATE_ROOM = 102;
  S_CREATE_ROOM = 103;
  C_JOIN_ROOM = 104;
  S_JOIN_ROOM = 105;
  C_GET_ROOM_LIST = 106;
  S_ROOM_LIST = 107;
  C_ENTER_LOBBY = 110;
  S_ENTER_LOBBY = 111;
  C_LEAVE_ROOM = 112;
  S_LEAVE_ROOM = 113;

  // Chat (120-129)
  C_CHAT = 120;
  S_CHAT = 121;

  // In-Game Logic (200-299)
  S_SPAWN_OBJECT = 200;
  S_DESPAWN_OBJECT = 201;
  S_MOVE_OBJECT_BATCH = 202;
  C_MOVE = 203;

  // Combat & Skills (300-399)
  C_USE_SKILL = 300;
  S_SKILL_EFFECT = 301;
  S_DAMAGE_EFFECT = 302;
  S_PLAYER_DOWNED = 303;
  S_PLAYER_REVIVE = 304;

  // Progression (400-499)
  S_EXP_CHANGE = 400;
  S_LEVEL_UP_OPTION = 401;
  C_SELECT_LEVEL_UP = 402;

  // System (500-599)
  S_GAME_WIN = 500;
  S_GAME_OVER = 501;
}

// ==========================================================
// Meta & Login
// ==========================================================

message C_Login {
  string username = 1;
  string password = 2;
  string config_file_path = 3; // Optional
}

message S_Login {
  bool success = 1;
  int32 my_player_id = 2;
  float map_width = 3;
  float map_height = 4;
}

message C_CreateRoom {
    int32 wave_pattern_id = 1;
}

message S_CreateRoom {
    bool success = 1;
    int32 room_id = 2;
}

message C_JoinRoom {
    int32 room_id = 1;
}

message S_JoinRoom {
    bool success = 1;
    int32 room_id = 2;
}

message C_EnterLobby {
}

message S_EnterLobby {
    bool success = 1;
}

message C_LeaveRoom {
}

message S_LeaveRoom {
    bool success = 1;
}

// Room Info
message RoomInfo {
    int32 room_id = 1;
    int32 current_players = 2;
    int32 max_players = 3;
    bool is_playing = 4;
}

message C_GetRoomList {
    bool only_joinable = 1;  // true면 입장 가능한 방만
}

message S_RoomList {
    repeated RoomInfo rooms = 1;
}

message C_Chat {
    string msg = 1;
}

message S_Chat {
    int32 player_id = 1;
    string msg = 2;
}

// ==========================================================
// In-Game Logic
// ==========================================================

enum ObjectType {
  UNKNOWN = 0;
  PLAYER = 1;
  MONSTER = 2;
  PROJECTILE = 3;
  ITEM = 4;
}

enum ObjectState {
    IDLE = 0;
    MOVING = 1;
    ATTACKING = 2;
    DEAD = 3;
    DOWNED = 4; // Player only
}

message ObjectInfo {
  int32 object_id = 1;
  ObjectType type = 2;
  int32 type_id = 3; // e.g., Monster ID or Class ID
  float x = 4;
  float y = 5;
  int32 hp = 6;
  int32 max_hp = 7;
  ObjectState state = 8;
  int32 owner_id = 9; // For projectiles
}

message S_SpawnObject {
  repeated ObjectInfo objects = 1;
}

message S_DespawnObject {
  repeated int32 object_ids = 1;
}

// Batched Movement Update
// This is critical for performance with 500+ entities
message ObjectPos {
  int32 object_id = 1;
  float x = 2;
  float y = 3;
  float vx = 4; // Velocity for client-side prediction
  float vy = 5;
}

message S_MoveObjectBatch {
  repeated ObjectPos moves = 1;
}

message C_Move {
  // Client only sends its own input or position
  // In Server-Authoritative, client sends input direction
  float dir_x = 1;
  float dir_y = 2;
}

// ==========================================================
// Combat & Skills
// ==========================================================

message C_UseSkill {
    int32 skill_id = 1;
    float target_x = 2;
    float target_y = 3;
}

message S_SkillEffect {
    int32 caster_id = 1;
    int32 skill_id = 2;
    float x = 3;
    float y = 4;
    repeated int32 target_ids = 5; // Who got hit
}

message S_DamageEffect {
    repeated int32 target_ids = 1;
    repeated int32 damage_values = 2;
}

message S_PlayerDowned {
    int32 player_id = 1;
}

message S_PlayerRevive {
    int32 player_id = 1;
}

// ==========================================================
// Progression
// ==========================================================

message S_ExpChange {
    int32 current_exp = 1;
    int32 max_exp = 2;
    int32 level = 3;
}

message LevelUpOption {
    int32 option_id = 1;
    int32 skill_id = 2;
    string name = 3;
    string desc = 4;
    bool is_new = 5; // true = new, false = upgrade
}

message S_LevelUpOption {
    repeated LevelUpOption options = 1;
}

message C_SelectLevelUp {
    int32 option_index = 1; // 0, 1, 2
}

message S_GameWin {
    int64 total_time_ms = 1;
    int32 kill_count = 2;
}

message S_GameOver {
    int64 survived_time_ms = 1;
}
